src/Map.php

```diff
@@ -399,14 +487,22 @@ public function pairs(): Sequence
      */
     public function put($key, $value)
     {
-        $pair = $this->lookupKey($key);
-
-        if ($pair) {
-            $pair->value = $value;
+        $pairRef = $this->lookupKey($key, $hash);
 
+        if ($pairRef) {
+            $pairRef->pair->value = $value;
         } else {
             $this->checkCapacity();
-            $this->pairs[] = new Pair($key, $value);
+            $pair = new Pair($key, $value);
+            $pairIndex = $this->nextPairIndex++;
+            $pairRef = new PairRef($pairIndex, $pair);
+            $this->pairRefs[$pairIndex] = $pairRef;
+
+            if (!array_key_exists($hash, $this->table)) {
+                $this->table[$hash] = [$pairRef];
+            } else {
+                $this->table[$hash][] = $pairRef;
```

Comment:
does this allow you to have multiple pairrefs associated with a hash?

Reply:
Yep! We have to handle collisions. We're only switching to an array when there's an actual collision in order to save memory (since arrays allocate a lot of memory and collisions should be rare).